src/Nethermind/Nethermind.Merge.Plugin/ProcessedTransactionsDbCleaner.cs

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Extensions;
+using Nethermind.Db;
+using Nethermind.Logging;
+
+namespace Nethermind.Merge.Plugin;
+
+public class ProcessedTransactionsDbCleaner : IDisposable
+{
+    private readonly IBlockFinalizationManager _finalizationManager;
+    private readonly IDb _processedTxsDb;
+    private readonly ILogger _logger;
+    private long _lastFinalizedBlock = 0;
+
+    public ProcessedTransactionsDbCleaner(IBlockFinalizationManager finalizationManager, IDb processedTxsDb, ILogManager logManager)
+    {
+        _finalizationManager = finalizationManager ?? throw new ArgumentNullException(nameof(finalizationManager));
+        _processedTxsDb = processedTxsDb ?? throw new ArgumentNullException(nameof(processedTxsDb));
+        _logger = logManager.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+
+        _finalizationManager.BlocksFinalized += OnBlocksFinalized;
+    }
+
+    private void OnBlocksFinalized(object? sender, FinalizeEventArgs e)
+    {
+        if (e.FinalizedBlocks.Count > 0 && e.FinalizedBlocks[0].Number > _lastFinalizedBlock)
```

Comment:
In scenario of the merge plugin this e.FinalizedBlocks won't be null, but something to consider if null check would be good here

Reply:
ok, resolved this